Using OrderedDict by default as some devices need ordering

dbarrosop opened the pull request:
I am using pyangbind to translate OpenConfig into native configuration and I am running into issues as some devices require ordering when entering IP addresses. For example, the following configuration will fail:

```
interface Ethernet2.1
    encapsulation dot1q vlan 1
    ip address 172.20.0.1/24 secondary
    ip address 192.168.1.1/24
napalm_base.exceptions.MergeConfigException: Error [1000]: CLI command 17 of 40 'ip address 172.20.0.1/24 secondary' failed: could not run command [Primary address must be assigned first]
```

The solution is as simple as using always `OrderedDict` instead of when only `user_ordered = True`.

dbarrosop commented:
BTW, I know the tests are failing. I will look into them but I wanted to first send the PR to see if this is something that you will want to merge or not.

robshakir commented:
Hi David,

Interesting one - I'm not sure I quite understand the issue here. Is the problem here in the ordering of which address is primary in the dict? I had two ways to address this previously:

- Don't make the primary address deterministic.
- Use an extension to indicate which the primary address is which is then used by the config generation template/code.

It'd be good to discuss wider, because there's some implications AFAICS of changing this from `dict`->`OrderedDict`, especially where folks are using `isinstance()`

Cheers!
r.

dbarrosop commented:
> Don't make the primary address deterministic.

That probably has two consequences:
1. You will get annoying and unnecessary diffs as the secondary keywords just shifts around.
2. It might also have an impact on some protocols as the primary IP might be used to form adjacencies or to select which network to advertise.

> Use an extension to indicate which the primary address is which is then used by the config generation template/code.

Not sure I understand, I wrote an augmentation to set that field and that's working fine. The problem is when a iterate over the list I don't have any way of ensuring I process the IP addresses in the same order I entered them unless I use an OrderedDict.

> It'd be good to discuss wider, because there's some implications AFAICS of changing this from dict->OrderedDict, especially where folks are using isinstance()

One would hope people is using `is_container|leaf` or `_yang_type` to do type checks but I understand, that's why I wanted to discuss first before looking into CI.

The thing is that I wrote an open source framework that can parse any native configuration/show_command and transform it into a YANG model and vice versa so treating this as an individual case (and potentially others) might be problematic. It would be much better if we could fix it in the source.

sjtarik commented:
To David's point, JUNOS RPC calls also expect the fields in parameters are in the order given by the Junos RPC Yang file.

robshakir commented:
@sjtarik - I think this is a different issue. This implies that we need to store somewhere what order we had the statements in from the file and then serialise this with that order. This is unsafe (since JSON objects' keys don't maintain their order).

The issue that @dbarrosop raises is actually related to lists AIUI, essentially, we'd like to always use an ordered dictionary such that the order of insertion matters.

@dbarrosop - let's try and go ahead and merge this, I think it seems OK, and we have had this issue elsewhere.

dbarrosop commented:
>  I think this is a different issue

yes, that's correct. Those are two different issues. The issue where fields have to be ordered (oh, the joy), I haven't yet figured out up to what extent it's really a problem. The only case I have found so far is that the `name` of the interface has to be the very first parameter when sending the object to the device but I am sure there will be others and, probably, on other vendors too.

> let's try and go ahead and merge this, I think it seems OK, and we have had this issue elsewhere.

Thanks! I will fix CI.

dbarrosop commented:
Ok, I think this is it. Quick notes, as you were using OrderedDict sometimes to know if the order was relevant, in order to replicate the old behavior I added a `_user_ordered` attribute to the OrderedDict object and used that instead. So, now, instead of `isinstance(d, OrderedDict)` we can do `getattr(d, "_user_ordered", False)` and achieve the same result. No need to change `isinstance(d, dict)` as `OrderedDict` evaluates `True`.

Oh, I also added `values` and `items` as those were missing.

robshakir commented:
awesome, thanks @dbarrosop! LGTM, and thanks for signing the Google CLA a while back. Will merge this now.
